File: src/kyth_shared/kyth_shared/desktop/windows_installer.py
"""Safe, testable Windows-installer workflow backed by Bottles."""
from __future__ import annotations

import hashlib
import json
import os
import re
import shutil
import stat
import subprocess
from collections.abc import Callable
from dataclasses import dataclass
from enum import StrEnum
from pathlib import Path
from typing import Any

from ..commands import APPLICATION_RUNNER, CommandRunner

BOTTLES_ID = "com.usebottles.bottles"
FLATHUB_URL = "https://dl.flathub.org/repo/flathub.flatpakrepo"


class InstallerKind(StrEnum):
    EXE = "exe"
    MSI = "msi"


class Compatibility(StrEnum):
    LIKELY = "likely"
    UNKNOWN = "unknown"
    UNSUPPORTED = "unsupported"


class WorkflowFailureKind(StrEnum):
    INVALID_FILE = "invalid-file"
    FILE_CHANGED = "file-changed"
    BOTTLES_INSTALL = "bottles-install"
    BOTTLE_CREATE = "bottle-create"
    FILE_STAGE = "file-stage"
    LAUNCH = "launch"


class WorkflowFailure(RuntimeError):
    def __init__(self, kind: WorkflowFailureKind, message: str) -> None:
        super().__init__(message)
        self.kind = kind


@dataclass(frozen=True)
class FileIdentity:
    device: int
    inode: int
    size: int
    modified_ns: int


@dataclass(frozen=True)
class InstallerRequest:
    source: Path
    kind: InstallerKind
    architecture: str
    identity: FileIdentity
    sha256: str


@dataclass(frozen=True)
class CompatibilityAssessment:
    level: Compatibility
    summary: str
    detail: str


@dataclass(frozen=True)
class BottlePlan:
    name: str
    environment: str
    architecture: str


@dataclass(frozen=True)
class StagedInstaller:
    host_path: Path
    sandbox_path: Path


@dataclass(frozen=True)
class LaunchResult:
    bottle: BottlePlan
    staged: StagedInstaller


ProgressCallback = Callable[[str], None]

_PE_MACHINES = {
    0x014C: "win32",
    0x8664: "win64",
    0xAA64: "arm64",
}
_UNSUPPORTED_PATTERNS = re.compile(
    r"(?:^|[-_. ])(?:anti[-_. ]?cheat|battleye|easyanti(?:cheat)?|driver|firmware|"
    r"bios|chipset|microsoft[-_. ]?store|windows[-_. ]?update)(?:$|[-_. ])",
    re.IGNORECASE,
)
_GAMING_PATTERNS = re.compile(
    r"(?:game|gaming|steam|battle[-_. ]?net|blizzard|gog|epic|launcher|ubisoft|uplay)",
    re.IGNORECASE,
)


def _identity(path: Path) -> FileIdentity:
    info = path.stat()
    return FileIdentity(info.st_dev, info.st_ino, info.st_size, info.st_mtime_ns)


def _sha256(path: Path) -> str:
    digest = hashlib.sha256()
    with path.open("rb") as source:
        for chunk in iter(lambda: source.read(1024 * 1024), b""):
            digest.update(chunk)
    return digest.hexdigest()


def _inspect_pe(source: Any) -> str:
    header = source.read(64)
    if len(header) < 64 or header[:2] != b"MZ":
        raise WorkflowFailure(
            WorkflowFailureKind.INVALID_FILE,
            "The file does not contain a valid Windows executable header.",
        )
    pe_offset = int.from_bytes(header[0x3C:0x40], "little")
    if pe_offset < 64 or pe_offset > 64 * 1024 * 1024:
        raise WorkflowFailure(
            WorkflowFailureKind.INVALID_FILE,
            "The Windows executable header points outside a safe inspection range.",
        )
    source.seek(pe_offset)
    pe_header = source.read(6)
    if len(pe_header) != 6 or pe_header[:4] != b"PE\0\0":
        raise WorkflowFailure(
            WorkflowFailureKind.INVALID_FILE,
            "The file has a DOS header but no valid PE executable header.",
        )
    return _PE_MACHINES.get(int.from_bytes(pe_header[4:6], "little"), "unknown")


def inspect_installer(filepath: str | os.PathLike[str]) -> InstallerRequest:
    """Validate an installer by content and capture an immutable launch identity."""
    path = Path(filepath).expanduser()
    try:
        if path.is_symlink() or not path.is_file():
            raise WorkflowFailure(
                WorkflowFailureKind.INVALID_FILE,
                "Choose a regular, non-symbolic-link installer file.",
            )
        resolved = path.resolve(strict=True)
        mode = resolved.stat().st_mode
        if not stat.S_ISREG(mode):
            raise WorkflowFailure(
                WorkflowFailureKind.INVALID_FILE,
                "Choose a regular installer file.",
            )
        suffix = resolved.suffix.lower()
        with resolved.open("rb") as source:
            if suffix == ".exe":
                kind = InstallerKind.EXE
                architecture = _inspect_pe(source)
            elif suffix == ".msi":
                if source.read(8) != b"\xd0\xcf\x11\xe0\xa1\xb1\x1a\xe1":
                    raise WorkflowFailure(
                        WorkflowFailureKind.INVALID_FILE,
                        "The file does not contain a valid MSI compound-document header.",
                    )
                kind = InstallerKind.MSI
                architecture = "win64"
            else:
                raise WorkflowFailure(
                    WorkflowFailureKind.INVALID_FILE,
                    "Kyth currently supports Windows .exe and .msi installers only.",
                )
        return InstallerRequest(
            source=resolved,
            kind=kind,
            architecture=architecture,
            identity=_identity(resolved),
            sha256=_sha256(resolved),
        )
    except WorkflowFailure:
        raise
    except OSError as exc:
        raise WorkflowFailure(
            WorkflowFailureKind.INVALID_FILE,
            f"The installer could not be read: {exc}",
        ) from exc


def assess_compatibility(request: InstallerRequest) -> CompatibilityAssessment:
    """Return an honest, conservative compatibility assessment."""
    if request.architecture == "arm64":
        return CompatibilityAssessment(
            Compatibility.UNSUPPORTED,
            "ARM Windows installer",
            "This installer targets Windows on ARM, which this Kyth compatibility path does not support.",
        )
    if _UNSUPPORTED_PATTERNS.search(request.source.stem):
        return CompatibilityAssessment(
            Compatibility.UNSUPPORTED,
            "System-level Windows component",
            "Drivers, firmware tools, kernel anti-cheat, and Windows system components generally cannot run through Wine.",
        )
    if request.architecture in {"win32", "win64"}:
        return CompatibilityAssessment(
            Compatibility.LIKELY,
            "Standard Windows installer",
            "Many conventional desktop installers work, but compatibility is not guaranteed.",
        )
    return CompatibilityAssessment(
        Compatibility.UNKNOWN,
        "Unknown Windows architecture",
        "Kyth can try this installer, but its architecture could not be identified reliably.",
    )


def plan_bottle(request: InstallerRequest) -> BottlePlan:
    stem = re.sub(r"[^a-z0-9]+", "-", request.source.stem.lower()).strip("-")
    for token in ("setup", "installer", "install", "update", "updater"):
        stem = re.sub(rf"(?:^|-){token}(?:-|$)", "-", stem).strip("-")
    stem = stem[:36] or "windows-app"
    suffix = request.sha256[:8]
    architecture = request.architecture if request.architecture in {"win32", "win64"} else "win64"
    environment = "gaming" if _GAMING_PATTERNS.search(request.source.stem) else "application"
    return BottlePlan(f"Kyth-{stem}-{suffix}", environment, architecture)


def flatpak_install_commands() -> tuple[list[str], list[str]]:
    return (
        ["flatpak", "remote-add", "--if-not-exists", "--user", "flathub", FLATHUB_URL],
        ["flatpak", "install", "-y", "--noninteractive", "--user", "flathub", BOTTLES_ID],
    )


def bottles_cli(*args: str) -> list[str]:
    return ["flatpak", "run", "--command=bottles-cli", BOTTLES_ID, *args]


def bottle_names(payload: str) -> set[str]:
    """Parse Bottles JSON across the list shapes used by supported releases."""
    try:
        data = json.loads(payload)
    except (json.JSONDecodeError, TypeError):
        return {line.strip() for line in payload.splitlines() if line.strip()}
    if isinstance(data, dict):
        data = data.get("bottles", data)
        if isinstance(data, dict):
            return {str(name) for name in data}
    if not isinstance(data, list):
        return set()
    names: set[str] = set()
    for item in data:
        if isinstance(item, str):
            names.add(item)
        elif isinstance(item, dict):
            name = item.get("Name") or item.get("name")
            if name:
                names.add(str(name))
    return names


def stage_installer(request: InstallerRequest, home: Path | None = None) -> StagedInstaller:
    """Copy the installer into Bottles' private Flatpak storage."""
    if _identity(request.source) != request.identity:
        raise WorkflowFailure(
            WorkflowFailureKind.FILE_CHANGED,
            "The installer changed after it was inspected. Reopen it to continue safely.",
        )
    user_home = (home or Path.home()).resolve()
    relative = Path("kyth-installers") / request.sha256[:16]
    host_dir = user_home / ".var" / "app" / BOTTLES_ID / "cache" / relative
    safe_name = re.sub(r"[^A-Za-z0-9._ -]", "_", request.source.name)
    host_path = host_dir / safe_name
    try:
        host_dir.mkdir(parents=True, exist_ok=True, mode=0o700)
        if not host_path.exists() or _sha256(host_path) != request.sha256:
            temporary = host_path.with_suffix(host_path.suffix + ".part")
            with request.source.open("rb") as source, temporary.open("wb") as target:
                shutil.copyfileobj(source, target, length=1024 * 1024)
            os.chmod(temporary, 0o600)
            if _sha256(temporary) != request.sha256:
                temporary.unlink(missing_ok=True)
                raise WorkflowFailure(
                    WorkflowFailureKind.FILE_CHANGED,
                    "The installer changed while it was being prepared. Reopen it to continue safely.",
                )
            temporary.replace(host_path)
        # Flatpak exposes its per-app cache at this same absolute host path.
        # Passing ~/.cache here would address the host's cache, which Bottles
        # cannot access without weakening its sandbox.
        return StagedInstaller(host_path, host_path)
    except WorkflowFailure:
        raise
    except OSError as exc:
        raise WorkflowFailure(
            WorkflowFailureKind.FILE_STAGE,
            f"Could not prepare the installer inside the Bottles sandbox: {exc}",
        ) from exc


class WindowsInstallerWorkflow:
    """Prepare Bottles and launch one validated installer without a shell."""

    def __init__(self, runner: CommandRunner | None = None, *, home: Path | None = None) -> None:
        self.runner = runner or APPLICATION_RUNNER
        self.home = home

    def _run(
        self,
        command: list[str],
        failure: WorkflowFailureKind,
        message: str,
        *,
        timeout: float = 120,
    ) -> str:
        try:
            result = self.runner.run(
                command, capture_output=True, text=True, timeout=timeout
            )
        except (OSError, subprocess.SubprocessError) as exc:
            raise WorkflowFailure(failure, f"{message}: {exc}") from exc
        if result.returncode != 0:
            detail = (result.stderr or result.stdout or "unknown error").strip()
            raise WorkflowFailure(failure, f"{message}: {detail}")
        return result.stdout or ""

    def bottles_installed(self) -> bool:
        try:
            result = self.runner.run(
                ["flatpak", "info", BOTTLES_ID], capture_output=True, text=True
            )
        except (OSError, subprocess.SubprocessError):
            return False
        return result.returncode == 0

    def ensure_bottles(self, progress: ProgressCallback) -> None:
        if self.bottles_installed():
            return
        progress("Preparing Flathub…")
        remote, install = flatpak_install_commands()
        self._run(remote, WorkflowFailureKind.BOTTLES_INSTALL, "Could not configure Flathub")
        progress("Installing Bottles…")
        self._run(
            install,
            WorkflowFailureKind.BOTTLES_INSTALL,
            "Could not install Bottles",
            timeout=3600,
        )

    def ensure_bottle(self, plan: BottlePlan, progress: ProgressCallback) -> None:
        progress("Checking isolated environments…")
        output = self._run(
            bottles_cli("--json", "list", "bottles"),
            WorkflowFailureKind.BOTTLE_CREATE,
            "Could not list Bottles environments",
        )
        if plan.name in bottle_names(output):
            return
        progress("Creating an isolated Windows environment…")
        self._run(
            bottles_cli(
                "new",
                "--bottle-name",
                plan.name,
                "--environment",
                plan.environment,
                "--arch",
                plan.architecture,
            ),
            WorkflowFailureKind.BOTTLE_CREATE,
            "Could not create the Windows environment",
            timeout=1800,
        )

    def execute(
        self,
        request: InstallerRequest,
        progress: ProgressCallback | None = None,
    ) -> LaunchResult:
        report = progress or (lambda _message: None)
        self.ensure_bottles(report)
        plan = plan_bottle(request)
        self.ensure_bottle(plan, report)
        report("Copying the installer into the sandbox…")
        staged = stage_installer(request, self.home)
        report("Opening the Windows installer…")
        try:
            self.runner.spawn(
                bottles_cli("run", "-b", plan.name, "-e", os.fspath(staged.sandbox_path)),
                start_new_session=True,
            )
        except (OSError, subprocess.SubprocessError) as exc:
            raise WorkflowFailure(
                WorkflowFailureKind.LAUNCH,
                f"Bottles could not launch the installer: {exc}",
            ) from exc
        return LaunchResult(plan, staged)

File: src/kyth-shared-rs/src/system/windows_installer.rs
//! Safe Windows-installer inspection and Bottles planning.
//!
//! This is the read-only half of `desktop.windows_installer`: it validates PE
//! and MSI headers, captures a file identity/hash, assesses compatibility, and
//! projects a deterministic bottle plan. Staging, Flatpak installation, and
//! launching a Windows program remain explicit caller-owned actions.

use regex::RegexBuilder;
use serde::{Deserialize, Serialize};
use serde_json::Value;
use sha2::{Digest, Sha256};
use std::collections::BTreeSet;
use std::fmt::{Display, Formatter};
use std::fs::File;
use std::io::{Read, Seek, SeekFrom};
use std::os::unix::fs::MetadataExt;
use std::path::{Path, PathBuf};

pub const BOTTLES_ID: &str = "com.usebottles.bottles";
pub const FLATHUB_URL: &str = "https://dl.flathub.org/repo/flathub.flatpakrepo";

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum InstallerKind { Exe, Msi }

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum Compatibility { Likely, Unknown, Unsupported }

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum WorkflowFailureKind { InvalidFile, FileChanged }

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct InstallerInspectionError {
    pub kind: WorkflowFailureKind,
    pub message: String,
}

impl Display for InstallerInspectionError {
    fn fmt(&self, formatter: &mut Formatter<'_>) -> std::fmt::Result { self.message.fmt(formatter) }
}

impl std::error::Error for InstallerInspectionError {}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct FileIdentity {
    pub device: u64,
    pub inode: u64,
    pub size: u64,
    pub modified_ns: i128,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct InstallerRequest {
    pub source: PathBuf,
    pub kind: InstallerKind,
    pub architecture: String,
    pub identity: FileIdentity,
    pub sha256: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct CompatibilityAssessment {
    pub level: Compatibility,
    pub summary: String,
    pub detail: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct BottlePlan {
    pub name: String,
    pub environment: String,
    pub architecture: String,
}

fn invalid(message: impl Into<String>) -> InstallerInspectionError {
    InstallerInspectionError { kind: WorkflowFailureKind::InvalidFile, message: message.into() }
}

fn inspect_pe(source: &mut File) -> Result<String, InstallerInspectionError> {
    let mut dos_header = [0_u8; 64];
    source.read_exact(&mut dos_header).map_err(|_| invalid("The file does not contain a valid Windows executable header."))?;
    if &dos_header[..2] != b"MZ" {
        return Err(invalid("The file does not contain a valid Windows executable header."));
    }
    let pe_offset = u32::from_le_bytes(dos_header[0x3c..0x40].try_into().unwrap()) as u64;
    if pe_offset < 64 || pe_offset > 64 * 1024 * 1024 {
        return Err(invalid("The Windows executable header points outside a safe inspection range."));
    }
    source.seek(SeekFrom::Start(pe_offset)).map_err(|_| invalid("The Windows executable header could not be inspected."))?;
    let mut header = [0_u8; 6];
    source.read_exact(&mut header).map_err(|_| invalid("The Windows executable header points outside a safe inspection range."))?;
    if &header[..4] != b"PE\0\0" {
        return Err(invalid("The file has a DOS header but no valid PE executable header."));
    }
    Ok(match u16::from_le_bytes([header[4], header[5]]) {
        0x014c => "win32",
        0x8664 => "win64",
        0xaa64 => "arm64",
        _ => "unknown",
    }.into())
}

fn identity(path: &Path) -> Result<FileIdentity, InstallerInspectionError> {
    let metadata = std::fs::metadata(path).map_err(|error| invalid(format!("The installer could not be read: {error}")))?;
    Ok(FileIdentity { device: metadata.dev(), inode: metadata.ino(), size: metadata.len(), modified_ns: i128::from(metadata.mtime()) * 1_000_000_000 + i128::from(metadata.mtime_nsec()) })
}

fn sha256(path: &Path) -> Result<String, InstallerInspectionError> {
    let mut source = File::open(path).map_err(|error| invalid(format!("The installer could not be read: {error}")))?;
    let mut digest = Sha256::new();
    let mut buffer = [0_u8; 1024 * 1024];
    loop {
        let count = source.read(&mut buffer).map_err(|error| invalid(format!("The installer could not be read: {error}")))?;
        if count == 0 { break; }
        digest.update(&buffer[..count]);
    }
    Ok(format!("{:x}", digest.finalize()))
}

pub fn inspect_installer(path: impl AsRef<Path>) -> Result<InstallerRequest, InstallerInspectionError> {
    let path = path.as_ref();
    if path.is_symlink() || !path.is_file() {
        return Err(invalid("Choose a regular, non-symbolic-link installer file."));
    }
    let resolved = path.canonicalize().map_err(|error| invalid(format!("The installer could not be read: {error}")))?;
    let kind = match resolved.extension().and_then(|extension| extension.to_str()).map(str::to_ascii_lowercase).as_deref() {
        Some("exe") => InstallerKind::Exe,
        Some("msi") => InstallerKind::Msi,
        _ => return Err(invalid("Kyth currently supports Windows .exe and .msi installers only.")),
    };
    let mut source = File::open(&resolved).map_err(|error| invalid(format!("The installer could not be read: {error}")))?;
    let architecture = match kind {
        InstallerKind::Exe => inspect_pe(&mut source)?,
        InstallerKind::Msi => {
            let mut header = [0_u8; 8];
            source.read_exact(&mut header).map_err(|_| invalid("The file does not contain a valid MSI compound-document header."))?;
            if header == *b"\xd0\xcf\x11\xe0\xa1\xb1\x1a\xe1" { "win64".into() } else { return Err(invalid("The file does not contain a valid MSI compound-document header.")); }
        }
    };
    Ok(InstallerRequest { source: resolved.clone(), kind, architecture, identity: identity(&resolved)?, sha256: sha256(&resolved)? })
}

pub fn assess_compatibility(request: &InstallerRequest) -> CompatibilityAssessment {
    let unsupported = RegexBuilder::new(r"(?:^|[-_. ])(?:anti[-_. ]?cheat|battleye|easyanti(?:cheat)?|driver|firmware|bios|chipset|microsoft[-_. ]?store|windows[-_. ]?update)(?:$|[-_. ])").case_insensitive(true).build().expect("static compatibility pattern");
    if request.architecture == "arm64" {
        return CompatibilityAssessment { level: Compatibility::Unsupported, summary: "ARM Windows installer".into(), detail: "This installer targets Windows on ARM, which this Kyth compatibility path does not support.".into() };
    }
    let stem = request.source.file_stem().and_then(|stem| stem.to_str()).unwrap_or_default();
    if unsupported.is_match(stem) {
        return CompatibilityAssessment { level: Compatibility::Unsupported, summary: "System-level Windows component".into(), detail: "Drivers, firmware tools, kernel anti-cheat, and Windows system components generally cannot run through Wine.".into() };
    }
    if matches!(request.architecture.as_str(), "win32" | "win64") {
        return CompatibilityAssessment { level: Compatibility::Likely, summary: "Standard Windows installer".into(), detail: "Many conventional desktop installers work, but compatibility is not guaranteed.".into() };
    }
    CompatibilityAssessment { level: Compatibility::Unknown, summary: "Unknown Windows architecture".into(), detail: "Kyth can try this installer, but its architecture could not be identified reliably.".into() }
}

pub fn plan_bottle(request: &InstallerRequest) -> BottlePlan {
    let source_stem = request.source.file_stem().and_then(|stem| stem.to_str()).unwrap_or("windows-app").to_ascii_lowercase();
    let separators = regex::Regex::new(r"[^a-z0-9]+").expect("static bottle name pattern");
    let mut stem = separators.replace_all(&source_stem, "-").trim_matches('-').to_string();
    for token in ["setup", "installer", "install", "update", "updater"] {
        let pattern = regex::Regex::new(&format!(r"(?:^|-){token}(?:-|$)")).expect("static bottle wrapper pattern");
        stem = pattern.replace_all(&stem, "-").trim_matches('-').to_string();
    }
    stem.truncate(36);
    let architecture = matches!(request.architecture.as_str(), "win32" | "win64").then_some(request.architecture.as_str()).unwrap_or("win64");
    let gaming = RegexBuilder::new(r"(?:game|gaming|steam|battle[-_. ]?net|blizzard|gog|epic|launcher|ubisoft|uplay)").case_insensitive(true).build().expect("static gaming pattern").is_match(&source_stem);
    BottlePlan { name: format!("Kyth-{}-{}", if stem.is_empty() { "windows-app" } else { &stem }, &request.sha256[..request.sha256.len().min(8)]), environment: if gaming { "gaming" } else { "application" }.into(), architecture: architecture.into() }
}

pub fn flatpak_install_commands() -> [Vec<String>; 2] {
    [vec!["flatpak", "remote-add", "--if-not-exists", "--user", "flathub", FLATHUB_URL].into_iter().map(String::from).collect(), vec!["flatpak", "install", "-y", "--noninteractive", "--user", "flathub", BOTTLES_ID].into_iter().map(String::from).collect()]
}

pub fn bottles_cli(args: &[&str]) -> Vec<String> {
    [vec!["flatpak", "run", "--command=bottles-cli", BOTTLES_ID].into_iter().map(String::from).collect::<Vec<_>>(), args.iter().map(|arg| (*arg).into()).collect()].concat()
}

pub fn bottle_names(payload: &str) -> BTreeSet<String> {
    let Ok(mut value) = serde_json::from_str::<Value>(payload) else {
        return payload.lines().map(str::trim).filter(|line| !line.is_empty()).map(String::from).collect();
    };
    if let Some(bottles) = value.get("bottles") { value = bottles.clone(); }
    if let Some(object) = value.as_object() { return object.keys().cloned().collect(); }
    value.as_array().into_iter().flatten().filter_map(|item| item.as_str().map(String::from).or_else(|| item.get("Name").or_else(|| item.get("name")).and_then(Value::as_str).map(String::from))).collect()
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::fs;
    use tempfile::tempdir;

    fn pe(machine: u16) -> Vec<u8> {
        let mut bytes = vec![0_u8; 128];
        bytes[..2].copy_from_slice(b"MZ");
        bytes[0x3c..0x40].copy_from_slice(&(64_u32).to_le_bytes());
        bytes[64..68].copy_from_slice(b"PE\0\0");
        bytes[68..70].copy_from_slice(&machine.to_le_bytes());
        bytes
    }

    #[test]
    fn inspects_pe_and_msi_headers() {
        let directory = tempdir().unwrap();
        let exe = directory.path().join("Setup Game.exe");
        fs::write(&exe, pe(0x8664)).unwrap();
        let request = inspect_installer(&exe).unwrap();
        assert_eq!(request.architecture, "win64");
        assert_eq!(assess_compatibility(&request).level, Compatibility::Likely);
        assert_eq!(plan_bottle(&request).environment, "gaming");
        let msi = directory.path().join("office.msi");
        fs::write(&msi, b"\xd0\xcf\x11\xe0\xa1\xb1\x1a\xe1payload").unwrap();
        assert_eq!(inspect_installer(&msi).unwrap().kind, InstallerKind::Msi);
    }

    #[test]
    fn rejects_bad_headers_and_system_components() {
        let directory = tempdir().unwrap();
        let exe = directory.path().join("driver.exe");
        fs::write(&exe, b"MZbad").unwrap();
        assert!(inspect_installer(&exe).is_err());
        let request = InstallerRequest { source: PathBuf::from("Battleye Setup.exe"), kind: InstallerKind::Exe, architecture: "win64".into(), identity: FileIdentity { device: 0, inode: 0, size: 0, modified_ns: 0 }, sha256: "0123456789abcdef".into() };
        assert_eq!(assess_compatibility(&request).level, Compatibility::Unsupported);
    }

    #[test]
    fn parses_bottles_shapes_and_projects_commands() {
        assert_eq!(bottle_names(r#"{"bottles":{"Demo":{}}}"#), BTreeSet::from(["Demo".into()]));
        assert_eq!(bottle_names(r#"[{"Name":"Demo"},"Other"]"#), BTreeSet::from(["Demo".into(), "Other".into()]));
        assert_eq!(bottle_names("Demo\nOther\n"), BTreeSet::from(["Demo".into(), "Other".into()]));
        assert_eq!(bottles_cli(&["list"]) [0], "flatpak");
        assert_eq!(flatpak_install_commands()[1].last().unwrap(), BOTTLES_ID);
    }
}
